jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java

```diff
@@ -0,0 +1,172 @@
+            }
+        });
+        HtmlContainerListenerRegistrationsCleaner propertyChangeCleaner = propertyChangeSubscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, propertyChangeCleaner));
+
+        Subscription itemChangeSubscription = dataContainer.addItemChangeListener(itemChangeEvent -> {
+            Object propertyValue = itemChangeEvent.getItem() != null
+                    ? EntitySystemAccess.getEntityEntry(itemChangeEvent.getItem()).getAttributeValue(property)
+                    : null;
+            updateHtmlContainerText(htmlContainer, formatter, propertyValue);
+        });
+        HtmlContainerListenerRegistrationsCleaner itemChangeCleaner = itemChangeSubscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, itemChangeCleaner));
+    }
+
+    @Override
+    public void bind(HtmlContainer htmlComponent, CollectionContainer<?> dataContainer) {
+        bind(htmlComponent, dataContainer, new HtmlContainerTextFormatter());
+    }
+
+    @Override
+    public void bind(HtmlContainer htmlContainer, CollectionContainer<?> dataContainer, Formatter<Object> formatter) {
+        Preconditions.checkNotNullArgument(htmlContainer);
+        Preconditions.checkNotNullArgument(dataContainer);
+        Preconditions.checkNotNullArgument(formatter);
+
+        updateHtmlContainerText(htmlContainer, formatter, dataContainer.getItems());
+
+        Subscription subscription = dataContainer.addCollectionChangeListener(collectionChangeEvent ->
+                updateHtmlContainerText(htmlContainer, formatter, dataContainer.getItems()));
+
+        HtmlContainerListenerRegistrationsCleaner htmlContainerListenerRegistrationsCleaner = subscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, htmlContainerListenerRegistrationsCleaner));
+    }
+
+    public void unbind(HtmlContainer htmlContainer) {
+        Preconditions.checkNotNullArgument(htmlContainer);
+
+        Iterator<Pair<HtmlContainer, HtmlContainerListenerRegistrationsCleaner>> iterator = listenersRegistrations.iterator();
+        while (iterator.hasNext()) {
+            Pair<HtmlContainer, HtmlContainerListenerRegistrationsCleaner> htmlContainerAndCleaner = iterator.next();
+            if (htmlContainer.equals(htmlContainerAndCleaner.getFirst())) {
+                htmlContainerAndCleaner.getSecond().remove();
+                iterator.remove();
+            }
+        }
+    }
+
+    protected void updateHtmlContainerText(HtmlContainer htmlContainer,
+                                           Formatter<Object> formatter,
+                                           @Nullable Object value) {
+        htmlContainer.setText(formatter.apply(value));
+    }
+
+    @FunctionalInterface
+    protected interface HtmlContainerListenerRegistrationsCleaner {
+        void remove();
+    }
+
+    class HtmlContainerTextFormatter implements Formatter<Object> {
```

Comment:
I don't like the concept of a non-static class that uses one `MetadataTools` bean from its parent
I think we should remove it

Moreover, the parent class doesn't use the bean at all O_O

You have a clear division between collection binding and instance binding
For collection binding, you can use the ready-made `io.jmix.flowui.component.formatter.CollectionFormatter`
And for a instance one, you can simply use a lambda with `MetadataTools#apply`